src/modules/playlist/components/PlaylistContainer.jsx

```diff
@@ -4,13 +4,36 @@ import WithPlaylistId from '../../common/components/WithPlaylistId';
 import React from 'react';
 import Playlist from './Playlist';
 import adapt from '../../common/components/Adapt';
+import PlaylistFragments from '../../common/fragments/PlaylistFragments';
 
 const TOGGLE_SEARCH = gql`
 	mutation ToggleSearch {
 		toggleSearch @client
 	}
 `;
 
+const CREATE_PLAYLIST = gql`
+	mutation CreatePlaylist($playlist: String!) {
+		upsertPlaylist(where: {name: $playlist}, create: {name: $playlist}, update: {}) {
+			...AllPlaylist
+		}
+	}
+
+	${PlaylistFragments.all}
+`;
+
+const createPlaylistUpdate = playlist => (cache, {data: {upsertPlaylist}}) => {
+	const query = gql`
+		query CreatePlaylistUpdate($playlist: String!) {
```

Comment:
Still new to this paradigm.. I'm a little confused about this being a query as its name sort of implies a create operation.

Reply:
Yeah this even confused me even when I was looking back at it; I'm still trying to work out the naming convention.

Here's basically what's going on: When the playlist mounted, say at: `localhost:8080/fake-room`, it would run a query like:
```graphql
query PlaylistQuery {
  playlist(where: {name: "fake-room"}) {
    id
    tracks {
      ...AllTrack
    }
}
```
And if that playlist didn't exist, we would just get `null` back, which Apollo would cache for that query:

![updateql1](https://user-images.githubusercontent.com/5115807/46106825-9559b080-c18e-11e8-8e45-6d190d6537b1.png)
Note that it actually caches the entire query.

The problem is, later, another component like `SearchResults` might run a mutation like:
```graphql
mutation AddTrack($playlist: String!) {
  createPlaylist(data: {name: $playlist, tracks: [...blah blah blah]}) {
    id
    tracks {
      ...AllTrack
    }
  }
}
```

Now we can look at this and realize that if we call it with `'fake-room'`, the playlists described by both of these operations are the same (because they have the same name and names are unique). But Apollo doesn't know that they are the same; the only way it can realize that that mutation updated the same playlist as the `PlaylistQuery` is if it could see that the `id`s were the same.

So we need to explicitly tell Apollo that the result from this mutation is also  result of the above query. Luckily this is pretty straightforward: we can pass an `update` callback function to our mutation that will take the return value of the mutation and the cache, and can update the cache. In code that might look like this:
```jsx
<SearchResult
  onClick={() => addTrackMutationFromAbove({
    variables: { playlist },
    update: (cache, mutationResult) => {
	// First we write a query for the data in the cache that we want to update
	const query = gql`
		query AddTrackUpdate($playlist: String!) {
			playlist(where: {name: $playlist}) {
				...AllPlaylist
			}
		}
	`;

	// Then we tell the cache to replace the data for for that query 
        // with the mutation result
	cache.writeQuery({
          query, 
          data: {playlist: mutationResult.data.addTrack}, 
          variables: {playlist}
        });
     }
  })  
/>
```
Because the query that we used to update the cache is the same as the query for the Playlist, Apollo will now realize that this result should be used for that query. And since it now has an ID, any further updates will work automatically (no cache fiddling)

So to answer your original question, why is there a query that seems to have a mutation name and do mutation things? The name structure I've been using is if it's used for a post mutation update, then the query name is `{MutationName}Update`. And it's tricky because it IS just a query, and just being used to read data. But then that query and some data are getting written into the store as sort of an extra-GraphQL mutation. 

Now finally, you might be wondering why the example I used is not how we do it in the app. And that brings us to the point of this PR. We USED to do it as above. But that meant that every operation that added something to the playlist need to 1) make sure that the playlist existed and then 2) run an update for that query after they made a mutation -- but only if that playlist didn't exist in the cache. Fragile, repetitive and easy to forget.

You can see in this code, we immediately create an empty playlist when the page [mounts](https://github.com/openrecord/vinyl/pull/63/files#diff-67d083b9c2f1e995c1a72a84d13b0100R9) that runs an [update](https://github.com/openrecord/vinyl/pull/63/files#diff-8c26a84eb3bb1975dbc2dc86ca4507d9R70) which will ensure the [cache](https://github.com/openrecord/vinyl/pull/63/files#diff-8c26a84eb3bb1975dbc2dc86ca4507d9R28) always has the playlist ID. Now every other mutation doesn't need to worry about updating the cache or creating the playlist. So long as they fetch the ID in their result, Apollo will notify all the queries that depend on it.